rmldatatfbbox/rmldatatfbbox/utils/helpers.py

```diff
@@ -0,0 +1,115 @@
+import os
+import json
+import cv2
+from pathlib import Path
+import tensorflow as tf
+from object_detection.utils import dataset_util
+
+### FUNCTIONS ###
+def construct_all(image_ids: list, temp_dir: Path):
+    labeled_images = {}
+    label_to_int_dict = {}
+
+    for image_id in image_ids:
+        labeled_images[image_id] = construct(image_id, temp_dir, label_to_int_dict)
+    
+    return labeled_images, label_to_int_dict
+
+def construct(image_id: str, temp_dir: Path, label_to_int_dict: dict):
+    if temp_dir is None:
+        cwd = Path.cwd()
+        data_dir = cwd / 'data'
+    else:
+        data_dir = temp_dir
+
+    image_filepath = None
+    image_type = None
+    file_extensions = [".png", ".jpg", ".jpeg"]
+    for extension in file_extensions:
+        if os.path.exists(data_dir / f'image_{image_id}{extension}'):
```

Comment:
Again, this is part of the more broad discussion about the interface between the core and the plugins. But I'm thinking instead of the core doing the `image_ids` discovery and therefore locking the filenames into a very specific format, it makes more sense to let each plugin glob for whatever data it needs in whatever way it wants and then pass lists of paths back to the core.

Reply:
I'm confused, this code is all in the plugin, not the core. Am I missing something? looks like the plugin is doing its own discovery here

Reply:
Sorry, this isn't the most relevant bit of code to comment on. But right now the core is responsible for discovering all of the image IDs. Based on my brief look, it does this by explicitly searching for `meta_*.json` files and extracting the IDs from there (see `data/helpers.py -> load_metadata`). The interactive filtering happens on these IDs before any plugin code is run. The plugin then gets the list of IDs and uses those to re-infer all the filepaths, which is what you see above.

The plugin should be at least able to override this behavior. Just continuing to spitball based on my earlier object-oriented proposal: perhaps there could be a DatasetWriter method along the lines of `load_ids` or something. It would take the path to the data and be expected to return a dict with IDs as keys. It would do the interactive filtering on these IDs, then pass along the dict to further methods. The default implementation could be similar to the existing one, where the IDs are inferred from `meta_*.json` and the dict values are just `(metadata_path, image_path)` tuples. However, if a plugin wanted to, it could override this method to do ID discovery however it wanted and store any arbitrary amount of information as the dict values.

Reply:
can you elaborate more on the use case for the generalized functionality @kevin3-black ? As long as we are using imagery generated by Starfish, we are certain each image will have a metadata file to hold labels and tags. Are you thinking of cases where image labels and/or tags may be stored in other file formats or naming schemes different from what we have defined in starfish?

Reply:
resolved from 7/22 meeting discussion